src/multihash.js

```diff
@@ -0,0 +1,361 @@
+     * are considered dirty and may contain garbage.
+     *
+     * @protected
+     */
+    this.buffer = new Uint8Array(IN_BYTES_PER_QUAD)
+
+    /**
+     * Offset is the number of bytes in we have written into the buffer. If
+     * offset is 0 it means that the buffer is effectively empty. When `offset`
+     * is equal to `this.buffer.length` we have a quad that can be processed.
+     *
+     * @protected
+     */
+    this.offset = 0
+
+    /**
+     * The layers of the tree. Each layer will contain either 0 or 1 nodes
+     * between writes. When we write into a hasher, if we have enough nodes
+     * leaves will be created and pushed into the `layers[0]` array, after
+     * which we flush and combining every two leaf into a node which is moved
+     * to the next layer. This process is repeated until we reach the top
+     * layer, leaving each layer either empty or with a single node.
+     *
+     * @type {Layers}
+     */
+    this.layers = [[]]
+  }
+
+  /**
+   * Return the total number of bytes written into the hasher. Calling
+   * {@link reset} will reset the hasher and the count will be reset to 0.
+   *
+   * @returns {bigint}
+   */
+  count() {
+    return this.bytesWritten
+  }
+
+  /**
+   * Computes the digest of all the data that has been written into this hasher.
+   * This method does not have side-effects, meaning that you can continue
+   * writing and call this method again to compute digest of all the data
+   * written from the very beginning.
+   */
+  digest() {
+    const buffer = new Uint8Array(MULTIHASH_SIZE)
+    this.digestInto(buffer)
+    return new Digest(buffer)
+  }
+
+  /**
+   * Computes the digest and writes into the given buffer. You can provide
+   * optional `byteOffset` to write digest at that offset in the buffer. By
+   * default the multihash prefix will be written into the buffer, but you can
+   * opt-out by passing `false` as the `asMultihash` argument.
+   *
+   * @param {Uint8Array} output
+   * @param {number} [byteOffset]
+   * @param {boolean} asMultihash
+   */
+  digestInto(output, byteOffset = 0, asMultihash = true) {
+    const { buffer, layers, offset } = this
+
+    // We do not want to mutate the layers, so we create a shallow copy of it
+    // which we will use to compute the root.
+    let [leaves, ...nodes] = layers
+
+    // If we have some bytes in the buffer we fill rest with zeros and compute
+    // leaves from it. Note that it is safe to mutate the buffer here as bytes
+    // past `offset` are considered dirty and should not be read.
+    if (offset > 0) {
+      leaves = [...leaves, ...split(pad(buffer.fill(0, offset)))]
+    }
+
+    const tree = build([leaves, ...nodes])
+    const height = tree.length - 1
+    const [root] = tree[height]
+
+    // Write the multihash prefix if requested
+    if (asMultihash) {
+      output.set(PREFIX, byteOffset)
+      byteOffset += PREFIX.length
+    }
```

Comment:
nit: I would have a new function to do so. Thoughts?
Otherwise, the `byteOffset` being optional is also tricky as it is required to use third param. Perhaps having an object for options?

Reply:
You mean have a separate function that encodes multihash prefix ? I mostly wanted to align it with `StreamingHasher` interface implemented by [blake3-multihash](https://github.com/web3-storage/blake3-multihash/blob/cd1c81c9376714a7ab3a125992f895d8b46e679b/src/type.ts#L48-L84)

That said you are right interface is bit clunky. I have created https://github.com/multiformats/js-multiformats/issues/260 so we could uplift the incremental hasher into multiformats. How about we iterate on the interface there and then we can reflect consensus here in a followup.